rust/src/operations/update.rs

```diff
@@ -0,0 +1,947 @@
+    },
+    prelude::SessionContext,
+};
+use datafusion_common::Result as DataFusionResult;
+use datafusion_common::{Column, DFSchema, ScalarValue};
+use datafusion_expr::{case, col, lit, when, Expr};
+use datafusion_physical_expr::{
+    create_physical_expr,
+    expressions::{self},
+    PhysicalExpr,
+};
+use futures::{future::BoxFuture, Stream, StreamExt};
+use parquet::file::properties::WriterProperties;
+use serde_json::{Map, Value};
+
+use crate::{
+    action::{Action, DeltaOperation, Remove},
+    delta_datafusion::{find_files, parquet_scan_from_actions, register_store},
+    storage::{DeltaObjectStore, ObjectStoreRef},
+    table_state::DeltaTableState,
+    DeltaResult, DeltaTable, DeltaTableError,
+};
+
+use super::{transaction::commit, write::write_execution_plan};
+
+/// Used to represent user input of either a Datafusion expression or string expression
+pub enum Expression {
+    /// Datafusion Expression
+    DataFusion(Expr),
+    /// String Expression
+    String(String),
+}
+
+impl From<Expr> for Expression {
+    fn from(val: Expr) -> Self {
+        Expression::DataFusion(val)
+    }
+}
+
+impl From<&str> for Expression {
+    fn from(val: &str) -> Self {
+        Expression::String(val.to_string())
+    }
+}
+impl From<String> for Expression {
+    fn from(val: String) -> Self {
+        Expression::String(val)
+    }
+}
+
+/// Updates records in the Delta Table.
+/// See this module's documentation for more information
+pub struct UpdateBuilder {
+    /// Which records to update
+    predicate: Option<Expression>,
+    /// How to update columns in a record that match the predicate
+    updates: HashMap<Column, Expression>,
+    /// A snapshot of the table's state
+    snapshot: DeltaTableState,
+    /// Delta object store for handling data files
+    object_store: Arc<DeltaObjectStore>,
+    /// Datafusion session state relevant for executing the input plan
+    state: Option<SessionState>,
+    /// Properties passed to underlying parquet writer for when files are rewritten
+    writer_properties: Option<WriterProperties>,
+    /// Additional metadata to be added to commit
+    app_metadata: Option<Map<String, serde_json::Value>>,
+    /// CastOptions determines how data types that do not match the underlying table are handled
+    /// By default an error is returned
+    cast_options: CastOptions,
+}
+
+#[derive(Default)]
+/// Metrics collected during the Update operation
+pub struct UpdateMetrics {
+    /// Number of files added.
+    pub num_added_files: usize,
+    /// Number of files removed.
+    pub num_removed_files: usize,
+    /// Number of rows updated.
+    pub num_updated_rows: usize,
+    /// Number of rows just copied over in the process of updating files.
+    pub num_copied_rows: usize,
+    /// Time taken to execute the entire operation.
+    pub execution_time_ms: u128,
+    /// Time taken to scan the files for matches.
+    pub scan_time_ms: u128,
```

Comment:
So the expected unit is actually ms since I'm just pulling the metrics [here](https://learn.microsoft.com/en-us/azure/databricks/delta/history#--operation-metrics-keys) 
I've changed the call sites to use `as_millis()` which returns a u128. Should I really explicitly downcast that to a u64?

Reply:
Yes that will always be safe on the timescales we care about.